python/shark_turbine/transforms/general/custom_op_expansion.py

```diff
@@ -0,0 +1,248 @@
+)
+
+
+class ExpandCustomOpsPass(Pass):
+    def __init__(
+        self, root_op: Operation, reg: dict[str, CustomOp] = ALL_CUSTOM_OP_REGS
+    ):
+        super().__init__(root_op)
+        self.reg = reg
+        # Track pending deletions in a dict to preserve order and unique.
+        self.ops_to_delete: dict[Operation, None] = {}
+        self.type_converter = NativeTypeConverter(root_op.context)
+        self.symbol_table = SymbolTable(root_op)
+
+    def delete_op(self, op):
+        self.ops_to_delete[op.operation] = None
+
+    def run(self):
+        for mr in self.funcs:
+            self.expand_func(mr.op)
+        for op in self.ops_to_delete.keys():
+            self.erase_unused_op(op)
+
+    def expand_func(self, func_op: Operation):
+        """Expands custom ops in a traced torch function.
+
+        This finds operations of the form:
+        %0 = torch.operator "torch.ns.op"
+
+        And looks them up in the ALL_CUSTOM_OP_REGS. If it originated from one of those
+        registered ops, then it will be expanded in place.
+        """
+        name_prefix = "torch."
+
+        for block in func_op.regions[0].blocks:
+            for op in block.operations:
+                if op.name == "torch.operator":
+                    custom_op_name = StringAttr(op.attributes["name"]).value
+                    if custom_op_name.startswith(name_prefix):
+                        local_name = custom_op_name[len(name_prefix) :]
+                        custom_op_reg = self.reg.get(local_name)
+                        if custom_op_reg is not None:
+                            self.expand_custom_op(custom_op_reg, op)
+
+    def expand_custom_op(self, op_reg: CustomOp, op: Operation):
+        original_operands: list[Value] = list(op.operands)
+        ksel = AOTKernelSelection(
+            op_reg, original_operands, list(op.results), self.type_converter
+        )
+        op_reg.select(ksel)
+
+        module_body = self.root_op.regions[0].blocks[0]
+        kb = InlineKernelBuilder(
+            ksel,
+            op,
+            type_converter=self.type_converter,
+            module_body=module_body,
+            symbol_table=self.symbol_table,
+        )
+        with kb.ip, kb.location:
+            op_reg.generate(ksel, kb)
+        assert kb.yielded, "Custom op generation did not yield_results()"
+
+        self.delete_op(op)
+
+
+class AOTKernelSelection(KernelSelection):
+    __slots__ = [
+        "operands",
+        "results",
+        "type_converter",
+    ]
+
+    def __init__(
+        self,
+        op: CustomOp,
+        operands: list[Value],
+        results: list[Value],
+        type_converter: NativeTypeConverter,
+    ):
+        super().__init__(op, len(operands))
+        self.operands = operands
+        self.results = results
+        self.type_converter = type_converter
+
+    def arg_tensor(self, arg: int) -> TensorArg:
+        # This is annoying: We have to go from the Torch MLIR type system to the
```

Comment:
. Can we not let the custom kernel tell us what the return Torch type will be? I'm not a huge fan of converting from torch-mlir type system.

Reply:
I can fiddle with it, but you'll have it anyway by virtue of the eager path through needing to construct meta tensors and all that. If this is something you ever need to see, we can work on it some more. There are already a number of places we need to flip flop like this...

Reply:
I guess if there is no easy solution, let's just do this for now.